### DISCUSSION TITLE
Does the rtx accel use rtx cores if available with fallback?

### DISCUSSION THREAD

**Discussion by trsh** [LuisaGroup/LuisaCompute#119] [Q&A] [answered]
I am checking out this example : https://github.com/LuisaGroup/LuisaCompute/blob/stable/src/tests/test_rtx.cpp

**Comment by Mike-Leo-Smith** [comment, marked as answer]
Mostly yes. But it's more like a driver thing, which we don't have explicit control. And we do observe that on WSL or with some Intel on-chip GPUs the OptiX/DXR support is missing.

**Reply by trsh** [reply]
Ok. So it might also just crash? Or you mean it can just fail to detect rtx hardware and do software ray tracing?

**Reply by swfly** [reply]
Yes it crashes when compiling shaders when your hardware does not support, from my limited experience

**Reply by Mike-Leo-Smith** [reply]
Yes

**Reply by trsh** [reply]
Can this be improved- Like querying for hardware capabilities? Getting back a flag `yes/no/maybe` and with `maybe` better fall back to software raytracing? Or similar?

**Reply by trsh** [reply]
Or some sort of error handling instead of crash.

**Reply by Mike-Leo-Smith** [reply]
I'm not sure if there is any API to check this and might need some time to figure it out. 

Anyway, the error handling should be improved, at least with some more specific error messages.

**Reply by Mike-Leo-Smith** [reply]
I have filed issue #120 (and sorry that we are preparing a paper these days so it might take longer to close the issue)

**Reply by trsh** [reply]
@Mike-Leo-Smith is there a wat to cut straight to Non-hwrt (i.e. not using RT cores) ray tracing on GPU?

**Reply by Mike-Leo-Smith** [reply]
I don't think there is. But I guess the equivalent solution is to implement a software one by yourself with Luisa's compute shaders?

**Reply by trsh** [reply]
Sure. I was just thinking maybe Luisa comes with some batteries for software ray tracing included :). 
Overall the process is not so easy- building/updating accel, tracing against it on gpu, etc.
